update Thorin

storywithoutend opened the pull request:
**Philosophical Changes**
- Created some interface types so that components that have similar properties share the same language. The biggest one is type WithColor which adds a "color" and "colorScheme" property. This is paired with utility functions for getting the right color in a theme. I took the function approach because when we move to Linaria we will need to translate property values into css module names and I saw this as a step towards that transition.
- For inputs, I changed the layout so that most of the related sub components are at the same level as the input. For example, the input and prefix component are on the same level where before the input had a wrapper div around it. Inputs are always the first child of the containing component which allows the use of sibling selectors to update the styling. The primary reason for this is because we have styling for when a component is showing the placeholder. Since we are using react-hook-form, the value of the input is not reactive, so I rely on css input:placeholder-shown to control the styling when in placeholder state.
- For Banner, I'm testing out using a type for icons instead of using ReactNode. Would like feedback if you think this is better.
- Reverted much of the changes that were made for yalc hot reloading support. It was causing problems when building. I think part of the problem was by linking direction to the js file instead of the folder, it removed access to the package.json file.
 
**Changes list**
- When publishing a branch that is not main, will release a version with tag @next. 
- Copy README.md to thorin package when published
- Updated README.md
- Added Banner component
- Update Button. Simplified the component to focus on the actual use cases that we have. Added the counter component as well.
- Updated Field. Updated the styling and placement. Simplified on actual use cases we have. 
- Added RecordItem
- Update Typography: Uses the design language that Dom uses in Figma. 
- Updated Checkbox: Removed the toggle option
- Added Currency Toggle: Is a modified checkbox
- Updated Dropdown: The styling of the menu.
- Update Input: Removed the number type and unit. Simplified the component to our use cases. Built in the clear button. 
- Updated Profile: Removed the caret icon to the right. This is not the final form that profile will have since Dom has not finalized the styling but it is getting it closer to where he wants it, which was primarily to get the size of the button to be the same as a large button so it wouldn't cause a resizing shift when loading the page.
- Updated Select. Updated the styling and simplified component. Switched to using the Field ids. This component currently needs some work where it doesn't properly change the placeholder state when using react-hook-form. The problem is that you can't update the value of the input natively using the inputRef since ref in react-hook-form is not a normal ref but a functional ref so calling inputRef.current return undefined.
- Updated Dialog. Moved step counter to footer. Added an alert option to Dialog heading.
- For other components, it was mostly just updating the styling

TateB commented on components/src/components/molecules/Dropdown/Dropdown.tsx:
```diff
@@ -540,6 +550,7 @@ export const Dropdown = ({
           return React.cloneElement(child as any, {
             ...buttonProps,
             zindex: '10',
+            pressed: isOpen,
```
not all custom elements will have a pressed attribute, this shows a warning in unit tests

storywithoutend replied:
I switched it to passing a string when isOpen is true since strings are valid values for injected properties.

TateB commented on components/src/components/molecules/Dropdown/Dropdown.tsx:
```diff
@@ -556,6 +567,7 @@ export const Dropdown = ({
         setIsOpen={setIsOpen}
         shortThrow={shortThrow}
         width={
+          inner &&
```
maybe have this be a prop like `inheritContentWidth` instead of disabling for all but inner dropdowns

storywithoutend replied:
done.

TateB commented on components/src/components/molecules/Input/Input.test.tsx:
```diff
@@ -45,65 +45,79 @@ describe('<Input />', () => {
-      const element = screen.getByLabelText(/funding/i)
-      expect(element).toHaveValue(null)
-    })
+  // describe('[type=number]', () => {
```
remove

TateB commented on components/src/components/molecules/PageButtons/PageButtons.tsx:
```diff
@@ -40,6 +40,16 @@ const PageButton = styled.button<{ $selected?: boolean; $size: Size }>(
     cursor: pointer;
+    font-size: ${theme.fontSizes.body};
+    line-height: ${theme.lineHeights.body};
+    font-weight: ${theme.fontWeights.bold};
```
should be medium weight

storywithoutend replied:
It's been changed to bold in figma > thorin > pagination

TateB replied:
just realised it's not the weight but the color, should be `greyPrimary` (from figma)

storywithoutend replied:
Ahh... yeah... fixed

TateB commented on components/src/tokens/border.ts:
```diff
@@ -27,4 +27,7 @@ export const radii = {
   '3xLarge': '24px',
   '4xLarge': '40px',
   full: '9999px',
+  input: '0.5rem',
+  card: '1rem',
+  checkbox: '0.25rem',
```
why are these custom values? also why `rem`?

storywithoutend replied:
I was going for a more descriptive field. Something where you wouldn't need to look at the size token in order to know which one to choose.

TateB commented on components/src/tokens/color.ts:
```diff
@@ -72,10 +72,14 @@ const categories = {
 } satisfies Record<string, CategoryItem>
 
 const gradients = {
+  accent:
+    'linear-gradient(330.4deg, #44BCF0 4.54%, #7298F8 59.2%, #A099FF 148.85%)',
```
i forgot to add the code for it, but the accent gradient should be derived in the `makeMode` function, not static

storywithoutend replied:
I'll leave this to you? I don't see how it could be computed since all the gradients seem different. I also added two more gradients that were in the thorin file but not in the color.ts

storywithoutend replied:
never mind... I figured out what you meant and added accent to gradients

TateB commented on components/src/types/index.ts:
```diff
@@ -38,3 +40,48 @@ export type OptionalTitle = AllOrNone<{
+  | 'label'
+  | 'labelHeading'
+
+export type Typography =
+  | 'Heading/H1'
+  | 'Heading/H2'
+  | 'Heading/H3'
+  | 'Heading/H4'
+  | 'Large/XL Normal'
+  | 'Large/XL Bold'
+  | 'Large/Normal'
+  | 'Large/Bold'
+  | 'Body/Normal'
+  | 'Body/Bold'
+  | 'Small/Normal'
+  | 'Small/Bold'
+  | 'Small/XS Normal'
+  | 'Small/XS Bold'
```
directly translating typography names from figma doesn't look right because it's not a normal naming scheme within js, maybe use dashes or camelCase?

storywithoutend replied:
went with camelCase

TateB commented on components/src/utils/getTypography.ts:
```diff
@@ -0,0 +1,87 @@
+  if (typography === 'Body/Normal') return theme.lineHeights.body
+  if (typography === 'Body/Bold') return theme.lineHeights.body
+  if (typography === 'Small/Normal') return theme.lineHeights.small
+  if (typography === 'Small/Bold') return theme.lineHeights.small
+  if (typography === 'Small/XS Normal') return theme.lineHeights.extraSmall
+  if (typography === 'Small/XS Bold') return theme.lineHeights.extraSmall
+  return '1rem'
+}
+
+const getFontWeight: TypographyFunc = (theme, typography) => {
+  if (!typography) return theme.fontWeights.normal
+  if (typography === 'small') return theme.fontWeights.normal
+  if (typography === 'large') return theme.fontWeights.normal
+  if (typography === 'extraLarge') return theme.fontWeights.normal
+  if (typography === 'label') return theme.fontWeights.normal
+  if (typography === 'labelHeading') return theme.fontWeights.normal
+  if (typography === 'Heading/H1') return theme.fontWeights.extraBold
+  if (typography === 'Heading/H2') return theme.fontWeights.bold
+  if (typography === 'Heading/H3') return theme.fontWeights.bold
+  if (typography === 'Heading/H4') return theme.fontWeights.bold
+  if (typography === 'Large/XL Normal') return theme.fontWeights.normal
+  if (typography === 'Large/XL Bold') return theme.fontWeights.bold
+  if (typography === 'Large/Normal') return theme.fontWeights.normal
+  if (typography === 'Large/Bold') return theme.fontWeights.bold
+  if (typography === 'Body/Normal') return theme.fontWeights.normal
+  if (typography === 'Body/Bold') return theme.fontWeights.bold
+  if (typography === 'Small/Normal') return theme.fontWeights.normal
+  if (typography === 'Small/Bold') return theme.fontWeights.bold
+  if (typography === 'Small/XS Normal') return theme.fontWeights.normal
+  if (typography === 'Small/XS Bold') return theme.fontWeights.bold
+  return '1rem'
+}
+
+export const getTypography: TypographyFunc = (theme, typography, field) => {
+  if (!typography) return 'initial'
+  if (field === 'fontSize') return getFontSize(theme, typography)
+  if (field === 'lineHeight') return getLineHeight(theme, typography)
+  if (field === 'fontWeight') return getFontWeight(theme, typography)
+  return 'initial'
+}
```
wouldn't these just be better as an object?

TateB replied:
meant something more along the lines of

```typescript
const fontDetails = {
  label: {
    size: theme.fontSizes.extraSmall,
    lineHeight: theme.lineHeights.extraSmall,
    weight: theme.fontWeights.normal,
  },
  // ... etc
}

const getFontSize = (typography) => fontDetails[typography].size
// ... etc
```

TateB reviewed (changes requested):
- imo `colorScheme` implies a different theme or set of colours, which is not what the prop sets. just a nit though
- quite a few legibility issues on various components from 9px text, extraSmall font size should be 12px
- idk if i broke it but `Hue` isn't correctly exported in the built types

banner:
- icon should be centred on desktop
- the ETH icon shouldn't be the default for info type, is just a placeholder in designs
- maybe add a subtle hover state?

recorditem:
- should be a hover state

input:
- when clicking the icon, the input isn't focused
- should probs keep `text` cursor across entire input for consistency
- clear icon animation doesn't look great imo, should probs just fade in faster without any scaling
- i think clearable should be false by default
- readOnly inputs should not be clearable

profile:
- why was the chevron removed?
- text is too far apart, i think adding `margin-top: -2px` to the top element looks better (this probably wasn't because of your changes, just comparing to figma)

select:
- options should have `cursor: pointer` across all content, instead of just in padding
- add gap/padding between options

storywithoutend commented on components/src/tokens/typography.ts:
```diff
@@ -1,26 +1,40 @@
-  root: '16px',
+  body: '1rem',
+  small: '0.875rem',
+  extraSmall: '0.5625rem',
```
Good catch. Thanks

storywithoutend commented:
> * imo `colorScheme` implies a different theme or set of colours, which is not what the prop sets. just a nit though

I'm open to a better term, looked through a list but didn't find a better one.

> * quite a few legibility issues on various components from 9px text, extraSmall font size should be 12px

I switched extraSmall to 12px font. Nice catch.

> * idk if i broke it but `Hue` isn't correctly exported in the built types

Hue wasn't being exported. Added it to exported types.

> 
> banner:
> 
> * icon should be centred on desktop
> * the ETH icon shouldn't be the default for info type, is just a placeholder in designs
> * maybe add a subtle hover state?
> 

Banner is a different component from Helper. 

I'll ask Dom about using the Eth icon as a default as well as the hover state. I believe he was working on hover colors already.

What do you think of using a type specific for svgs for the icons?

> recorditem:
> 
> * should be a hover state

Asking Dom about hover states.

> 
> input:
> 
> * when clicking the icon, the input isn't focused

Nice. I turned off pointer-events in icons so the click goes through to the input.

> * should probs keep `text` cursor across entire input for consistency

Done.

> * clear icon animation doesn't look great imo, should probs just fade in faster without any scaling

I decrease transition duration to 0.1s and increased the starting scale to 0.8. I thought it looked better but I'm willing to go with no scaling as well.

> * i think clearable should be false by default

After thinking about I agree.

> * readOnly inputs should not be clearable

Added a readOnly state where clear/action button does not show and cursor remains default.

> 
> profile:
> 
> * why was the chevron removed?

Dom has an update to the profile in the Navigation  where the chevron is removed. Also the chevron svg has been resized so it works for inputs and doesn't look right in Profile. Could be done using resizing though. I didn't follow Dom's design completely because he hasn't finalized the updated profile component.

> * text is too far apart, i think adding `margin-top: -2px` to the top element looks better (this probably wasn't because of your changes, just comparing to figma)

I agree that it does look a little funky. I added a negative 0.125rem margin in between.  I'll ask Dom for clarification as well.

> 
> select:
> 
> * options should have `cursor: pointer` across all content, instead of just in padding

Fixed. Nice catch.

> * add gap/padding between options

Added 2px padding in between.

storywithoutend commented on components/src/components/atoms/Typography/Typography.tsx:
```diff
@@ -111,30 +78,29 @@ type Props = {
-      variant,
       className,
-      weight,
+      typography = 'Body/Normal',
```
switched to fontVariant.

storywithoutend commented on components/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@ensdomains/thorin",
-  "version": "0.6.21",
+  "version": "0.6.22-alpha-5",
```
Will do.

TateB commented:
> I'm open to a better term, looked through a list but didn't find a better one.
> 
> >

maybe `colorStyle`? could also remove the prop entirely and instead allow the `color` prop to take a color instead of just a hue (e.g. `color="blue", colorScheme="primary"` -> `color="bluePrimary"`)

> Hue wasn't being exported. Added it to exported types.

this didn't fix the issue, created #88 to fix

> Banner is a different component from Helper.

meant vertically centred, not horizontally. sorry

> What do you think of using a type specific for svgs for the icons?

works well

> I decrease transition duration to 0.1s and increased the starting scale to 0.8. I thought it looked better but I'm willing to go with no scaling as well.

looks a lot better

> Dom has an update to the profile in the Navigation where the chevron is removed. 

that's fine then

storywithoutend commented:
> > I'm open to a better term, looked through a list but didn't find a better one.
> > >
> 
> maybe `colorStyle`? could also remove the prop entirely and instead allow the `color` prop to take a color instead of just a hue (e.g. `color="blue", colorScheme="primary"` -> `color="bluePrimary"`)
> 

I broke it into two interfaces, `withColor` which is for things like Typography where there is just one color needed and `withColorStyle` for things like buttons that have a group of colors.


> > Banner is a different component from Helper.
> 
> meant vertically centred, not horizontally. sorry
> 

You got a sharp eye. Fixed.
